src/algebra/direct_sum/internal.lean

```diff
@@ -167,3 +167,62 @@ direct_sum.to_algebra S _ (λ i, (A i).subtype) rfl (λ _ _ _ _, rfl) (λ _, rfl
   (A : ι → submodule S R) [h : set_like.graded_monoid A] (i : ι) (x : A i) :
   direct_sum.submodule_coe_alg_hom A (direct_sum.of (λ i, A i) i x) = x :=
 direct_sum.to_semiring_of _ rfl (λ _ _ _ _, rfl) _ _
+
+section add_submonoid
+
+open_locale direct_sum big_operators
+
+variables [add_comm_monoid ι] [semiring R]
+variables (𝒜 : ι → add_submonoid R)
+
+lemma direct_sum.coe_mul_apply_add_submonoid [set_like.graded_monoid 𝒜]
+  [Π (i : ι) (x : 𝒜 i), decidable (x ≠ 0)] (r r' : ⨁ i, 𝒜 i) (i : ι) :
+  ((r * r') i : R) =
+    ∑ ij in finset.filter (λ ij : ι × ι, ij.1 + ij.2 = i) (r.support.product r'.support),
+      r ij.1 * r' ij.2 :=
+begin
+  rw [direct_sum.mul_eq_sum_support_ghas_mul, dfinsupp.finset_sum_apply,
+    add_submonoid.coe_finset_sum],
+  simp_rw [direct_sum.coe_of_add_submonoid_apply, ←finset.sum_filter, set_like.coe_ghas_mul],
+end
+
+end add_submonoid
+
+section submodule
+
+open_locale direct_sum big_operators
+
+variables [add_comm_monoid ι] [comm_semiring R] [ring S] [algebra R S]
+variables (𝒜 : ι → submodule R S)
+
+lemma direct_sum.coe_mul_apply_submodule [set_like.graded_monoid 𝒜]
+  [Π (i : ι) (x : 𝒜 i), decidable (x ≠ 0)] (r r' : ⨁ i, 𝒜 i) (i : ι) :
```

Comment:
Can you switch to using the same variable names as we do in `direct_sum.submodule_coe_alg_hom_of`?